include/JSObjectProxy.hh

```diff
@@ -0,0 +1,192 @@
+  /**
+   * @brief Helper function for various JSObjectProxy methods, sets a key-value pair on a JSObject given a python string key and a JS::Value value
+   *
+   * @param jsObject - The underlying backing store JSObject for the JSObjectProxy
+   * @param key - The key to be assigned or deleted
+   * @param value - The JS::Value to be assigned
+   */
+  static void JSObjectProxy_set_helper(JS::HandleObject jsObject, PyObject *key, JS::HandleValue value);
+
+  /**
+   * @brief Comparison method (.tp_richcompare), returns appropriate boolean given a comparison operator and other pyobject
+   *
+   * @param self - The JSObjectProxy
+   * @param other - Any other PyObject
+   * @param op - Which boolean operator is being performed (Py_EQ for equality, Py_NE for inequality, all other operators are not implemented)
+   * @return PyObject* - True or false depending on result of comparison
+   */
+  static PyObject *JSObjectProxy_richcompare(JSObjectProxy *self, PyObject *other, int op);
+
+  /**
+   * @brief Helper function for JSObjectProxy_richcompare
+   *
+   * @param self - The PyObject on the left side of the operator (guaranteed to be a JSObjectProxy *)
+   * @param other - The PyObject on the right side of the operator
+   * @param visited
+   * @return bool - Whether the compared objects are equal or not
+   */
+  static bool JSObjectProxy_richcompare_helper(JSObjectProxy *self, PyObject *other, std::unordered_map<PyObject *, PyObject *> &visited);
+};
+
+
+/**
+ * @brief Struct for the methods that define the Mapping protocol
+ *
+ */
+static PyMappingMethods JSObjectProxy_mapping_methods = {
+  .mp_length = (lenfunc)JSObjectProxyMethodDefinitions::JSObjectProxy_length,
+  .mp_subscript = (binaryfunc)JSObjectProxyMethodDefinitions::JSObjectProxy_get,
+  .mp_ass_subscript = (objobjargproc)JSObjectProxyMethodDefinitions::JSObjectProxy_assign
+};
+
+/**
+ * @brief Struct for the JSObjectProxyType, used by all JSObjectProxy objects
+ *
+ */
+static PyTypeObject JSObjectProxyType = {
+  .ob_base = {{1, &PyType_Type}, 0},
+  .tp_name = "pythonmonkey.JSObjectProxy",
+  .tp_basicsize = sizeof(JSObjectProxy),
+  .tp_itemsize = 0,
+  .tp_dealloc = (destructor)JSObjectProxyMethodDefinitions::JSObjectProxy_dealloc,
+  .tp_vectorcall_offset = 0,
+  .tp_getattr = NULL,
+  .tp_setattr = NULL,
+  .tp_as_async = NULL,
+  .tp_repr = NULL,
+  .tp_as_number = NULL,
+  .tp_as_sequence = NULL,
+  .tp_as_mapping = &JSObjectProxy_mapping_methods,
+  .tp_hash = NULL,
+  .tp_call = NULL,
+  .tp_str = NULL,
+  .tp_getattro = (getattrofunc)JSObjectProxyMethodDefinitions::JSObjectProxy_get,
+  .tp_setattro = (setattrofunc)JSObjectProxyMethodDefinitions::JSObjectProxy_assign,
+  .tp_as_buffer = NULL,
+  .tp_flags = Py_TPFLAGS_DEFAULT
+  // | Py_TPFLAGS_DICT_SUBCLASS  // https://docs.python.org/3/c-api/typeobj.html#Py_TPFLAGS_DICT_SUBCLASS
+  // | Py_TPFLAGS_HAVE_GC     // @TODO (Caleb Aikens) need to figure out how to make GC work cross-language
+  | Py_TPFLAGS_MAPPING,
```

Comment:
https://github.com/Distributive-Network/PythonMonkey/actions/runs/5340441182/jobs/9680264892#step:7:651
`Py_TPFLAGS_MAPPING` does [not exist in Python < 3.10](https://docs.python.org/3/c-api/typeobj.html#Py_TPFLAGS_MAPPING)